api/http/management_filetransfer.go

```diff
@@ -0,0 +1,736 @@
+						params.SessionID, params.UserID, deviceTopic,
+						wsft.MessageTypeACK, nil,
+						latestOffset); err != nil {
+						return
+					}
+					return
+				}
+
+				// verify the offset property
+				propOffset, _ := msg.Header.Properties[PropertyOffset].(int64)
+				if propOffset != latestOffset {
+					responseError = errors.Wrap(errFileTranserFailed,
+						"wrong offset received")
+					return
+				}
+				latestOffset += int64(len(msg.Body))
+
+				_, err := c.Writer.Write(msg.Body)
+				if err != nil {
+					return
+				}
+
+				numberOfChunks++
+				if numberOfChunks >= ackSlidingWindowSend {
+					if err := h.publishFileTransferProtoMessage(
+						params.SessionID, params.UserID, deviceTopic,
+						wsft.MessageTypeACK, nil,
+						latestOffset); err != nil {
+						return
+					}
+					numberOfChunks = 0
+				}
+			}
+
+		// send a Ping message to keep the session alive
+		case <-ticker.C:
+			h.publishFileTransferPing(params.SessionID, deviceTopic)
+
+		// no message after timeout expired, stop here
+		case <-timeout.C:
+			responseError = errFileTranserTimeout
+			return
+		}
+	}
+}
+
+func (h ManagementController) DownloadFile(c *gin.Context) {
+	params, statusCode, err := h.getFileTransferParams(c)
+	if err != nil {
+		c.JSON(statusCode, gin.H{"error": err.Error()})
+		return
+	}
+
+	request := &model.DownloadFileRequest{}
+	if err := c.ShouldBindJSON(request); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{
+			"error": errors.Wrap(err, "invalid request body").Error(),
+		})
+		return
+	}
+
+	if err := request.Validate(); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{
+			"error": errors.Wrap(err, "bad request").Error(),
+		})
+		return
+	}
+
+	allowed, err := h.fileTransferAllowed(c, params.TenantID, params.Device.ID)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{
+			"error": errors.Wrap(err, "failed to check RBAC"),
+		})
+		return
+	} else if !allowed {
+		c.JSON(http.StatusForbidden, gin.H{
+			"error": "Access denied (RBAC).",
+		})
+		return
+	}
+
+	if err := h.app.DownloadFile(c, params.UserID, params.Device.ID,
+		*request.Path); err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{
+			"error": errors.Wrap(err, "bad request").Error(),
+		})
+		return
+	}
+
+	h.downloadFileResponse(c, params, request)
+}
+
+func (h ManagementController) uploadFileResponseHandleInboundMessages(c *gin.Context,
+	msgChan chan *natsio.Msg, errorChan chan error, latestAckOffset *int64,
+	latestAckOffsets chan int64) {
+	for {
+		select {
+		case wsMessage := <-msgChan:
+			msg, msgBody, err := h.decodeFileTransferProtoMessage(
+				wsMessage.Data)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+
+			// process incoming messages from the device by type
+			switch msg.Header.MsgType {
+
+			// error message, stop here
+			case wsft.MessageTypeError:
+				errorMsg := msgBody.(*wsft.Error)
+				errorChan <- errors.New(*errorMsg.Error)
+				return
+
```

Comment:
```suggestion

			case ws.MessageTypePing:
				// send pong
			case ws.MessageTypePong:
				// update timeout
```

Reply:
I'm not seeing ping/pong being handled for upload.

Reply:
I handled the ping messages, generating pong. I'm not sure it makes sense to do the "pings" when uploading, as we are already supposed to send a lot of messages and wait for acks. What do you think?

Reply:
I'm pretty sure don't need it for either upload nor download since there is a constant flow of messages in both directions. So this is good enough for me.